src/blueprint.zig

```diff
@@ -0,0 +1,254 @@
+const std = @import("std");
+
+const main = @import("main.zig");
+const Compression = main.utils.Compression;
+const ZonElement = @import("zon.zig").ZonElement;
+const vec = main.vec;
+const Vec3i = vec.Vec3i;
+
+const Array3D = main.utils.Array3D;
+const Block = main.blocks.Block;
+const NeverFailingAllocator = main.heap.NeverFailingAllocator;
+const User = main.server.User;
+
+pub const GameIdToBlueprintIdMapType = std.AutoHashMap(u16, u16);
+const BlockIdSizeType = u32;
+const BlockStorageType = u32;
+
+const BinaryWriter = main.utils.BinaryWriter;
+const BinaryReader = main.utils.BinaryReader;
+
+pub const blueprintVersion = 0;
+
+pub const BlueprintCompression = enum(u16) {
+	deflate,
+};
+
+pub const Blueprint = struct {
+	blocks: Array3D(Block),
+
+	pub fn init(allocator: NeverFailingAllocator) Blueprint {
+		return .{.blocks = .init(allocator, 0, 0, 0)};
+	}
+	pub fn deinit(self: *Blueprint, allocator: NeverFailingAllocator) void {
+		self.blocks.deinit(allocator);
+	}
+	pub fn clone(self: *Blueprint, allocator: NeverFailingAllocator) Blueprint {
+		var new = Blueprint.init(allocator);
+		new.blocks = self.blocks.clone(allocator);
+		return new;
+	}
+	const CaptureResult = union(enum) {
+		success: Blueprint,
+		failure: struct {x: i32, y: i32, z: i32, message: []const u8},
+	};
+
+	pub fn capture(allocator: NeverFailingAllocator, pos1: Vec3i, pos2: Vec3i) CaptureResult {
+		var self = Blueprint{.blocks = undefined};
```

Comment:
self isn't used before `self.blocks` is assigned, and it is never modified afterwards, so why not declare it below as a const?